### PR TITLE
fix(#748): route 8 RDM sites through _normalise_rdm, and fix its own gauge leak

### LINKED ISSUE #748: The #725 symmetrise-before-trace-normalise bug survives in 8 more RDM sites (ipeps_rdm, ipeps_excitations, honeycomb)
Follow-up to #725/#742, which fixed the split-CTM and PESS energy paths (PR #745). The same latent gauge bug is still present in three other modules, deliberately left out of that PR to keep a verified fix from growing an unverified tail.

## The bug

Every CTM environment tensor carries an arbitrary complex phase, so a raw RDM is defined only up to an overall complex scalar. With `R = e^{iφ}(H + K)`:

```
Herm(e^{iφ}(H + K)) = cos(φ)·H + sin(φ)·iK
```

Symmetrising before trace-normalising rescales the physical part by `cos φ` and mixes the anti-Hermitian part in with weight `sin φ`. Trace-normalising first cancels the scalar exactly. `_normalise_rdm` in `_ctm_tensor_energy.py:31-57` carries the full derivation.

## Sites (all confirmed symmetrise-then-normalise)

| file | lines |
|---|---|
| `src/tenax/algorithms/ipeps_rdm.py` | 98, 167, 247, 295 |
| `src/tenax/algorithms/ipeps_excitations.py` | 248, 285 |
| `src/tenax/algorithms/_ctm_honeycomb_energy.py` | 203, 268 |

Each is the two-line `rdm_mat = 0.5 * (rdm_mat + rdm_mat.conj().T)` followed by a trace divide, i.e. the exact pattern fixed in #745.

## Fix

Route through `_normalise_rdm`, as PR #745 did. Note the one non-obvious case from that PR: where a trace-floor guard sits between the two operations, the guard must read the **raw** trace — `tr Herm(R) = Re(tr R)` is not gauge-invariant, so a phase near π/2 drives it through zero while `|tr R|` stays O(1) and fires a fallback on a healthy RDM.

## Verification this needs

Real-valued states are bit-for-bit unchanged (`tr Herm(R) = tr R`), so this is **only observable on complex states** — which is exactly why it survived this long. The pattern that catches it is `tests/test_split_ctm_energy_gauge.py` (PR #745): phase each environment tensor independently and **include π/2**, since generic angles pass at ~1e-14 either way. On the split path that test went 13-failed → 14-passed across the fix; a sweep of generic angles alone would have been green on the bug.

Each of these three modules will need its own complex-state gauge test — none has one today.

Related: #721, #724, #725, #742, PR #745.

> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

### PR DESCRIPTION
**Pull request by yingjerkao**
> 🤖 **AI-generated PR** — written by Claude Code, posted by @yingjerkao.

Closes #748, and fixes a second bug found while verifying it that the issue did not anticipate.

## 1. The eight sites (what #748 asked for)

`ipeps_rdm.py` (4), `ipeps_excitations.py` (2) and `_ctm_honeycomb_energy.py` (2) all symmetrised the RDM *before* trace-normalising. With `R = e^{iφ}(H + K)`:

```
Herm(e^{iφ}(H + K)) = cos(φ)·H + sin(φ)·iK
```

so an environment phase — a gauge fixed by nothing in the sweep — rescales the physical part and mixes in the anti-Hermitian part. All eight now route through `_normalise_rdm`; **zero** occurrences of the old pattern remain.

## 2. `_normalise_rdm`'s own guard was gauge-dependent (not in the issue)

Applying the issue's prescription was **not sufficient** — a π/2 phase still moved the energy by **6.96e-03**. Rather than accept that, I measured each stage:

- The raw RDM is *exactly* a phase multiple under an env phase: `|ratio| = 1.00000000`, argument spread **0.0**. So the network is fine and normalisation is the only step that can break invariance.
- `|tr(ρ_raw)| = 3.47e-14` against `EPS = 1e-15` — only 35× apart.

`mat / (trace + EPS)` uses an **additive absolute** epsilon, and `e^{iφ}·tr + EPS ≠ e^{iφ}·(tr + EPS)`. The predicted error `EPS/|tr| ≈ 2.9e-2` matched the observed relative shift of **2.8e-2**. The guard was defeating the invariance it exists to protect.

This is not a corner case: the raw RDM's scale is *arbitrary* because environment tensors are renormalised every sweep, so a small `|tr|` is the normal situation. **It also affects the already-merged #745 split-CTM and PESS paths**, which share this helper.

Replaced with a relative, phase-preserving floor: `norm(mat)` is gauge-*invariant*, so the comparison is too, and dividing by `tr` itself keeps covariance exact whenever the RDM carries any trace.

## 3. A regression I introduced and caught

The first version of that floor, `EPS·‖mat‖`, **vanishes for a zero matrix** — turning `0/0` into NaN. It broke two tests: the excitation norm at `B = 0` (a legitimate input — a zero excitation vector must have zero norm) and an iPEPS run test. The floor is now kept strictly positive. Recorded here because it shows this helper is load-bearing well beyond the paths #748 lists.

## 4. Separate bug found on the way: `ctm()` could not run on a complex state

`prev_sv` was seeded `dtype=env.C1.dtype` — complex for a complex state — while the loop body assigns `_dense_svd(...)`, always real. `lax.while_loop` requires an invariant carry type, so it raised outright:

```
The input carry component carry[1] has type complex128[8] but the
corresponding output carry component has type float64[8]
```

Real states are unaffected (`C1.dtype` is already real), which is why nothing caught it. Fixed and pinned.

## Testing

Three new gauge suites, one per module as #748 requires, all **mutation-verified** rather than assumed:

| suite | tests | killed by reverting the fix |
|---|---|---|
| `test_ipeps_rdm_gauge` | 19 | **17** (independently for *both* mutants: the routing, and the floor) |
| `test_ipeps_excitations_gauge` | 33 | **32** |
| `test_ctm_honeycomb_energy_gauge` | 28 | **16** |

The honeycomb survivors are structurally vacuous rather than weak: `_rdm1` contracts `C*`/`L*` but not `R*`, so phasing an `R` tensor genuinely cannot change it.

### A fixture I built and threw away

My first fixture used a random **complex** site tensor, following the issue's "only observable on complex states". `ctm()` drives that to a degenerate environment with `|tr| ~ 1e-28` — far below the `EPS` floor — so *nothing* is invariant there and the test measured fixture pathology rather than the bug. (Standing lesson: CTM on random or near-product input is not a valid oracle.)

The replacement uses a **real** physical simple-update state, which exposes the bug *more* sharply: for a real symmetric raw RDM `M`, `Herm(iM) = i(M − Mᵀ)/2` vanishes outright at π/2. The issue's "only on complex states" framing is true of comparing two *unphased* runs; a gauge test applies the phase itself.

Refs #748, #725, #742, #745.

---

## 5. Review round: the zero-RDM branch was differentiable in value only (bc51ef9)

Codex flagged that the new floor left `_normalise_rdm` non-differentiable at an all-zero RDM. Verified, and it is worse than a corner case — it is a regression **this PR** introduced. `main` used `mat / (trace + EPS)`, whose cotangent at zero is a finite `1/EPS`; the relative floor fixed the value and replaced the gradient with NaN.

The trap is a JAX one: `jnp.linalg.norm` has derivative `x/‖x‖ = 0/0` at zero, and `jnp.where` does **not** short-circuit a VJP — the unselected branch still contributes `0 · NaN`.

Reachable on the production path, not merely in principle:

```python
jax.grad(lambda B: _compute_norm(A, B, env, k, d).real)(jnp.zeros_like(A))   # -> NaN
```

`_build_effective_matrices` builds the excitation `H_eff`/`N` from exactly that call under `vmap`, so `B = 0` — the input the floor was added to support — poisons the quasiparticle spectrum.

**Fix:** form the squared norm as a polynomial and fence `sqrt` off from zero with a second `where` on its *argument*. The value is Frobenius either way, so §1's gauge invariance is untouched (all 94 gauge tests still pass).

**Tests:** 8 new, each mutation-verified — and each mutation kills a *different* subset, so nothing in the file is decoration:

| mutation | tests killed |
|---|---|
| revert the `sqrt` fence (back to `jnp.linalg.norm`) | 3 |
| floor loses its strictly-positive fence | 5 |
| floor fallback replaced by zero | 6 |
| `stop_gradient` on the denominator | 2 (the finite-difference guards) |

## 6. These suites were not running in required CI

Found while adding the above: none of the three gauge files were registered in `conftest._FILE_MARKERS`, so they carried no marker and `-m core` — the **required** gate — deselected all 80. They ran only on push to `main` or behind the `run-full-tests` label. Registered as `core`; they cost ~14 s together, so the guards now actually guard.

## Out of scope, filed separately

The same `jnp.where`-does-not-fix-a-VJP idiom exists at four **pre-existing** sites (`_ad_primitives.py:104`, `_ctm_root_implicit_asym.py:344`, `_ctm_root_implicit_symmetric.py:393`, `_ctm_tensor_projector_2x2.py:53` — the last on the production-default phase gauge). Reproduced, but not fixed here: #789.

### CODE REVIEW

File: src/tenax/algorithms/_ctm_tensor_energy.py [hunk 1]
```diff
@@ -52,8 +52,34 @@ def _normalise_rdm(mat: jax.Array) -> jax.Array:
...
+
+    The floor is kept strictly positive so that an *identically zero* matrix
+    still returns zero rather than ``0/0``.  A relative floor alone vanishes
+    with the matrix, which turned the excitation norm at ``B = 0`` into NaN --
+    a legitimate input, since a zero excitation vector must have zero norm.
     """
-    mat = mat / (jnp.trace(mat) + EPS)
+    tr = jnp.trace(mat)
+    scale = jnp.linalg.norm(mat)
+    floor = EPS * jnp.where(scale > 0, scale, 1.0)
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/_ctm_tensor_energy.py]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Confirmed and fixed in bc51ef9.

Verified the claim directly rather than taking it on the description — `jax.grad` at the all-zero matrix returned `[nan nan nan nan]` for both real and complex inputs, while the forward value was correctly `0`. And it is a regression **this PR** introduced: `main` used `mat / (trace + EPS)`, whose cotangent at zero is a finite `1/EPS`. The relative floor fixed the value and broke the gradient.

Also confirmed reachable on the production path, not only in principle:

```python
jax.grad(lambda B: _compute_norm(A, B, env, k, d).real)(jnp.zeros_like(A))  # -> NaN
```

`_build_effective_matrices` builds the excitation `H_eff`/`N` from exactly that call under `vmap`, so `B = 0` — the input the floor exists to support — would have poisoned the quasiparticle spectrum.

**Fix:** form the squared norm as a polynomial (differentiable everywhere) and fence `sqrt` off from zero with a second `where` on its *argument*:

```python
sq = jnp.sum(mat.real**2 + mat.imag**2)
nonzero = sq > 0
scale = jnp.where(nonzero, jnp.sqrt(jnp.where(nonzero, sq, 1.0)), 0.0)
floor = EPS * jnp.where(nonzero, scale, 1.0)
```

The value is Frobenius either way, so the gauge invariance this helper exists for is untouched — all 94 gauge tests still pass.

**Tests:** 8 new, each mutation-verified rather than assumed. Reverting the fence kills 3; dropping the floor's strict positivity kills 5; replacing the floor fallback with zero kills 6; a `stop_gradient` on the denominator kills the 2 finite-difference guards. Each mutation kills a different subset, so no test in the file is decoration.

One thing the report understated: the same idiom appears at four **pre-existing** sites (`_ad_primitives.py:104`, `_ctm_root_implicit_asym.py:344`, `_ctm_root_implicit_symmetric.py:393`, `_ctm_tensor_projector_2x2.py:53`, the last on the production-default phase gauge). Reproduced there too — filed as #789 rather than folded into this branch.
